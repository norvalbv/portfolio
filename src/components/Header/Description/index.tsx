import React, { ReactElement } from 'react';
import classNames from 'utils/classNames';

export type DescriptionProps = {
  children?: string | JSX.Element;
  className?: string;
  testId?: string;
};

const Description = ({ children, className, testId }: DescriptionProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (typeof children === 'string') {
    return (
      <p
        className={classNames(className || 'truncate text-body text-neutral-N00')}
        data-testid={testId || `${children} Description`}
      >
        {children}
      </p>
    );
  }

  return children;
};

export default Description;
