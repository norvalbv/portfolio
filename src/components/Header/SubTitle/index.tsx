import React, { ReactElement } from 'react';
import classNames from 'utils/classNames';

export type SubtitleProps = {
  children?: string | JSX.Element;
  className?: string;
  testId?: string;
};

const Subtitle = ({ children, className, testId }: SubtitleProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (typeof children === 'string') {
    return (
      <h4
        className={classNames(className || 'truncate text-body text-neutral-N300')}
        data-testid={testId || `${children} Subtitle`}
      >
        {children}
      </h4>
    );
  }

  return children;
};

export default Subtitle;
