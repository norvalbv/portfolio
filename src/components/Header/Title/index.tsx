import React, { ReactElement } from 'react';
import classNames from 'utils/classNames';

export type TitleProps = {
  children?: string | JSX.Element;
  className?: string;
  testId?: string;
  /**
   * Default 3.
   */
  level?: 1 | 2 | 3;
};

const Title = ({ children, className, testId, level = 2 }: TitleProps): ReactElement | null => {
  if (!children) {
    return null;
  }

  if (typeof children !== 'string') {
    return children;
  }

  const titleTestId = testId || `${children} Title`;

  // Level 1
  if (level === 1) {
    return (
      <h1
        className={classNames(
          className,
          'mb-20 text-5xl font-semibold underline underline-offset-8 md:text-8xl'
        )}
        data-testid={titleTestId}
      >
        {children}
      </h1>
    );
  }

  // Level 2. Default
  if (level === 2) {
    return (
      <h2
        className={classNames(
          className,
          'mb-10 text-4xl font-semibold underline underline-offset-8 md:mb-20 md:text-6xl'
        )}
        data-testid={titleTestId}
      >
        {children}
      </h2>
    );
  }

  // Level 3
  return (
    <h3
      className={classNames(className, 'mb-10 text-2xl font-semibold  md:text-4xl')}
      data-testid={titleTestId}
    >
      {children}
    </h3>
  );
};

export default Title;
