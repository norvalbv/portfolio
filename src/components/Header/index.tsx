import React, { ReactElement } from 'react';
import classNames from 'utils/classNames';
import Title from 'components/Header/Title';
import Subtitle from 'components/Header/SubTitle';
import Description from './Description';

export type HeaderProps = {
  // Own props
  className?: string;

  // Title props
  title?: string | JSX.Element;
  titleTestId?: string;
  titleClassName?: string;

  // Subtitle props
  subtitle?: string | JSX.Element;
  subtitleTestId?: string;
  subtitleClassName?: string;

  // Description props
  description?: string | JSX.Element;
  descriptionTestId?: string;
  descriptionClassName?: string;
};

const Header = ({
  className,
  description,
  descriptionClassName,
  descriptionTestId,
  subtitle,
  subtitleClassName,
  subtitleTestId,
  title,
  titleClassName,
  titleTestId,
}: HeaderProps): ReactElement | null => {
  if (!title && !subtitle) {
    return null;
  }

  return (
    <header className={classNames(className || 'flex items-center justify-between')}>
      <section className="flex w-full flex-col gap-1">
        <Title testId={titleTestId} className={titleClassName}>
          {title}
        </Title>
        <Subtitle testId={subtitleTestId} className={subtitleClassName}>
          {subtitle}
        </Subtitle>
        <Description testId={descriptionTestId} className={descriptionClassName}>
          {description}
        </Description>
      </section>
    </header>
  );
};

export default Header;
