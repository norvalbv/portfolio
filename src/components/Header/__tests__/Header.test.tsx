import React, { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Header, { HeaderProps } from '..';

const HeaderWithAllProps: FC<Required<HeaderProps>> = Header;

describe('<Header />', () => {
  test('should render nothing if title and subtitle are not provided', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with title provided', () => {
    const tree = renderer.create(<Header title="Test title" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with subtitle provided', () => {
    const tree = renderer.create(<Header subtitle="Test subtitle" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render correctly with all props', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <HeaderWithAllProps
            link={{ to: '/', title: 'Test Title' }}
            className="grid grid-cols-1 gap-1"
            title="Test title"
            titleTestId="Custom title test id"
            titleClassName="text-h2"
            subtitle="Test subtitle"
            subtitleTestId="Custom subtitle test id"
            subtitleClassName="text-sm"
            description="Test Description"
            descriptionTestId="Test Description Id"
            descriptionClassName="text-neutral-N00"
          />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
