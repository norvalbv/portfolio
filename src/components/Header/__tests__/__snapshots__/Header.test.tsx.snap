// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render correctly with all props 1`] = `
<header
  className="grid grid-cols-1 gap-1"
>
  <section
    className="flex w-full flex-col gap-1"
  >
    <h3
      className="text-h2"
      data-testid="Custom title test id"
    >
      Test title
    </h3>
    <h4
      className="text-sm"
      data-testid="Custom subtitle test id"
    >
      Test subtitle
    </h4>
    <p
      className="text-neutral-N00"
      data-testid="Test Description Id"
    >
      Test Description
    </p>
  </section>
  <a
    className="justify-center group relative flex h-6 items-center gap-1 transition ease-productive text-primary-P500 hover:text-primary-P700"
    data-testid="Test Title Link"
    href="/"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="inline-block select-none text-sm"
    >
      Test Title
    </span>
    <span>
      <svg
        className="text-inherit"
        color="#58585C"
        fill="none"
        height={16}
        viewBox="0 0 16 16"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 8C9.6 9.4 5 14 5 14"
          stroke="url(#0506cbdda73c368)"
          strokeLinecap="round"
          strokeMiterlimit="10"
          strokeWidth="1.5"
        />
        <path
          d="M5 2C5 2 9.6 6.6 11 8"
          stroke="currentColor"
          strokeLinecap="round"
          strokeMiterlimit="10"
          strokeWidth="1.5"
        />
        <defs>
          <linearGradient
            gradientUnits="userSpaceOnUse"
            id="0506cbdda73c368"
            x1="4.5"
            x2="11.5"
            y1="11"
            y2="11"
          >
            <stop
              stopColor="currentColor"
              stopOpacity="0"
            />
            <stop
              offset="0.7"
              stopColor="currentColor"
            />
          </linearGradient>
        </defs>
      </svg>
    </span>
  </a>
</header>
`;

exports[`<Header /> should render correctly with subtitle provided 1`] = `
<header
  className="flex items-center justify-between"
>
  <section
    className="flex w-full flex-col gap-1"
  >
    <h4
      className="truncate text-body text-neutral-N300"
      data-testid="Test subtitle Subtitle"
    >
      Test subtitle
    </h4>
  </section>
</header>
`;

exports[`<Header /> should render correctly with title provided 1`] = `
<header
  className="flex items-center justify-between"
>
  <section
    className="flex w-full flex-col gap-1"
  >
    <h3
      className="truncate text-h3"
      data-testid="Test title Title"
    >
      Test title
    </h3>
  </section>
</header>
`;

exports[`<Header /> should render nothing if title and subtitle are not provided 1`] = `null`;
